### PR TITLE
Unify showSDocUnsafe

### PR DESCRIPTION
**Pull request by July541**
We choose `printOutputable :: Outputable a => a -> Text` finally.

---

Extends #2828, this pr wrapped `showSDocUnsafe` and  unified all `SDoc` printing functions into the following four functions
1. `showGhc` has `defaultUserStyle` and `suppressUnique`, we'd use this as much as we can to print ghc internal.
2. `showGhcWithUniques` is the alias of `showSDocUnsafe`, it may have different behavior under different ghc versions, like #2716.
3. `prettyPrint` sames as `showGhc` but return `String` instead of `Text`.
4. `prettyPrintWithUnique`, the `String` version of `showGhcWithUniques`.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2830"></a>

### CODE REVIEW

File: ghcide/src/Development/IDE/GHC/Util.hs [hunk 1]
```diff
@@ -318,4 +314,45 @@ traceAst lbl x
...
+#endif
 
+-- | Print a GHC value by default `showSDocUnsafe`.
+--
+-- You may prefer `prettyPrint` unless you know what you are doing.
+--
+-- It internal using `unsafeGlobalDynFlags`.
+--
+-- `String` version of `showGhcWithUniques`.
+prettyPrintWithUniques :: Outputable a => a -> String
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/GHC/Util.hs:331]: I think basically nobody should ever want to do this, right? there's only one use, so maybe inline it there?

File: ghcide/src/Development/IDE/GHC/Util.hs [hunk 2]
```diff
@@ -318,4 +314,45 @@ traceAst lbl x
...
+-- `String` version of `showGhc`.
+prettyPrint :: Outputable a => a -> String
+prettyPrint = printWithoutUniques
+
+-- | Print a GHC value by default `showSDocUnsafe`.
+--
+-- You may prefer `showGhc` unless you know what you are doing.
+--
+-- It internal using `unsafeGlobalDynFlags`.
+showGhcWithUniques :: Outputable a => a -> T.Text
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/GHC/Util.hs:349]: this is unused, maybe let's not even expose it

File: ghcide/src/Development/IDE/GHC/Compat/Outputable.hs [hunk 3]
```diff
@@ -68,14 +67,20 @@ import qualified Outputable                      as Out
 import           SrcLoc
 #endif
 
-printNameWithoutUniques :: Outputable a => a -> String
-printNameWithoutUniques =
+printWithoutUniques :: Outputable a => a -> String
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/GHC/Compat/Outputable.hs:70]: haddock?

File: ghcide/src/Development/IDE/GHC/Util.hs [hunk 4]
```diff
@@ -318,4 +314,45 @@ traceAst lbl x
...
+showGhcWithUniques = T.pack . showSDocUnsafe . ppr
+
+-- | Print a GHC value in `defaultUserStyle` without unique symbols.
+--
+-- This is the most common print utility, will print with a user friendly style like: `a_a4ME` as `a`.
+--
+-- It internal using `unsafeGlobalDynFlags`.
+showGhc :: Outputable a => a -> T.Text
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/GHC/Util.hs:357]: In the interests of keeping things simple, I think we could get away with just this function. Replacing `prettyPrint` just requires `show $ showGhc x` at the call site.

Also, name bikeshedding: how about `printOutputable`?

File: ghcide/src/Development/IDE/LSP/Outline.hs [hunk 5]
```diff
@@ -250,10 +251,7 @@ defDocumentSymbol l = DocumentSymbol { .. } where
   _tags           = Nothing
 
 showRdrName :: RdrName -> Text
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/LSP/Outline.hs:253]: inline?

File: ghcide/src/Development/IDE/Types/Exports.hs [hunk 6]
```diff
@@ -178,7 +178,7 @@ unpackAvail mn
   | otherwise = const []
   where
     !mod = pack $ moduleNameString mn
-    f id@IdentInfo {..} = (pack (prettyPrint name), moduleNameText,[id])
+    f id@IdentInfo {..} = (pack (occNameString name), moduleNameText,[id])
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Types/Exports.hs:181]: why different here?

**Review by michaelpj — COMMENTED** [review]
Wonderful, thank you!

**Comment by isovector** [comment]
lgtm, thanks!

File: ghcide/src/Development/IDE/GHC/Compat/Outputable.hs [hunk 7]
```diff
@@ -91,14 +96,9 @@ printSDocQualifiedUnsafe unqual doc =
     showSDocForUser unsafeGlobalDynFlags unqual doc
 #endif
 
-printSDocAllTheWay :: DynFlags -> SDoc -> String
+
 #if MIN_VERSION_ghc(9,2,0)
-printSDocAllTheWay dflags sdoc = renderWithContext ctxt sdoc
```
Comment (Ailrun) [inline ghcide/src/Development/IDE/GHC/Compat/Outputable.hs:96]: Why this `#if` is still here? I mean, we can use the negated condition, cannot we?

**Review by Ailrun — COMMENTED** [review]
Thank you for the update!

File: ghcide/src/Development/IDE/LSP/Outline.hs (hunk 5 above)
Comment (July541) [reply ghcide/src/Development/IDE/LSP/Outline.hs:253]: I'm just a little worried about whether we're too radical. It turns out we also have some `occNameString` and I keep it for semantic clarity, or you're thinking to rule them all with`printOutputable`?

File: ghcide/src/Development/IDE/Types/Exports.hs (hunk 6 above)
Comment (July541) [reply ghcide/src/Development/IDE/Types/Exports.hs:181]: The same one, just a lonely clear name.

File: ghcide/src/Development/IDE/GHC/Util.hs (hunk 4 above)
Comment (July541) [reply ghcide/src/Development/IDE/GHC/Util.hs]: I keeped `printOutputable :: Outputable a => a -> String` and `printOutputableText :: Outputable a => a -> Text`. I think we shouldn't pay for extra `O(n)` to `pack`  if `Text` is not required, and also, a lot of `pack $ printOutputable x` looks inconvenient.

It is only my personal aspect, I'm happy to revise them if you have different views:)

File: ghcide/src/Development/IDE/Spans/AtPoint.hs [hunk 8]
```diff
@@ -55,6 +55,7 @@ import           Data.Version                         (showVersion)
 import           Development.IDE.Types.Shake          (WithHieDb)
 import           HieDb                                hiding (pointCommand)
 import           System.Directory                     (doesFileExist)
+import Development.IDE.GHC.Util (printOutputableText)
```
Comment (Ailrun) [inline ghcide/src/Development/IDE/Spans/AtPoint.hs:59]: Here is some ill-formatted code

File: ghcide/src/Development/IDE/GHC/Compat/Outputable.hs [hunk 9]
```diff
@@ -91,24 +100,15 @@ printSDocQualifiedUnsafe unqual doc =
...
 oldRenderWithStyle dflags sdoc sty = Out.renderWithStyle (initSDocContext dflags sty) sdoc
 oldMkUserStyle _ = Out.mkUserStyle
 oldMkErrStyle _ = Out.mkErrStyle
 
 oldFormatErrDoc :: DynFlags -> Err.ErrDoc -> Out.SDoc
 oldFormatErrDoc dflags = Err.formatErrDoc dummySDocContext
   where dummySDocContext = initSDocContext dflags Out.defaultUserStyle
-
-#else
+#elif !MIN_VERSION_ghc(9,2,0)
```
Comment (Ailrun) [inline ghcide/src/Development/IDE/GHC/Compat/Outputable.hs:111]: How about `(9,0,0)`? I think that's more clear.

**Review by Ailrun — APPROVED** [review]
Thanks, looks good to me :) Here are a few very minor comments:

File: ghcide/src/Development/IDE/GHC/Util.hs (hunk 4 above)
Comment (michaelpj) [reply ghcide/src/Development/IDE/GHC/Util.hs]: I was thinking the other way around: just have the one that prints `Text` and let people `show` it if they need a `String`. `Text` by default is a better way anyway!

I think the cost of string conversions can be assumed to be totally trivial given the size of things we're talking about! I'm more concerned with the complexity of the codebase, so the fewer functions the better...

File: ghcide/src/Development/IDE/LSP/Outline.hs (hunk 5 above)
Comment (michaelpj) [reply ghcide/src/Development/IDE/LSP/Outline.hs]: I don't know - I don't know how this code works. I just noticed that it's trivial so maybe it should be inlined.

File: ghcide/src/Development/IDE/Types/Exports.hs (hunk 6 above)
Comment (michaelpj) [reply ghcide/src/Development/IDE/Types/Exports.hs:181]: I have no idea if this is sensible or not, I don't know this bit of the code! Just wondering. Probably could do with a comment.

Comment (July541) [reply ghcide/src/Development/IDE/Types/Exports.hs:181]: Replaced with `printOutputable`

**Review by eddiemundo — APPROVED** [review]
I wanna bikeshed `printOutputable` to `prettyPrintAsText` since it's the same as `prettyPrint`, or oppositely `prettyPrint` to `printOutputableAsString`, but since you already did a change once, it's fine...

So all good.

**Comment by michaelpj** [comment]
Temporarily taking out of the merge queue until the release is done.

**Pull request merged**